File: py-dexter/dexter/equilibrium/currents.py
"""Defines the different plasma current objects as wrappers over `_PyCurrent`."""

import numpy as np
from semver import Version

from dexter._core import _PyCurrent

from .utils import LastClosedFluxSurface
from .base import EquilibriumObject, Current
from dexter.types import Array1, FluxCoordinateState, Interpolation1dType, NetCDFVersion


class LarCurrent(EquilibriumObject, Current):
    r"""Analytical Large Aspect Ratio Current with $g=1$ and $I=0$.

    !!! note

        No $\psi/\psi_p$ bounds checks are performed in evaluations.

    Example
    -------
    ``` py
    >>> current = dex.LarCurrent()

    ```
    """

    _r: _PyCurrent

    def __init__(self) -> None:
        self._r = _PyCurrent.build_lar()
        super(EquilibriumObject, self).__init__()
        super(Current, self).__init__()


class NcCurrent(EquilibriumObject, Current):
    r"""Numerical plasma current profile reconstructed from a netCDF file.

    Related quantities are computed by interpolating over the data arrays.

    Parameters
    ----------
    path
        The path to the NetCDF file.
    interp_type
        The 1D interpolation type.

    Example
    -------
    ``` py
    >>> current = dex.NcCurrent(path, "Cubic")

    ```
    """

    _r: _PyCurrent

    def __init__(self, path: str, interp_type: Interpolation1dType) -> None:
        self._r = _PyCurrent.build_nc(path, interp_type)
        super(EquilibriumObject, self).__init__()
        super(Current, self).__init__()

    @property
    def path(self) -> str:
        """The path to the NetCDF file."""
        return self._r.path

    @property
    def netcdf_version(self) -> NetCDFVersion:
        """The path to the NetCDF file."""
        return Version.parse(self._r.netcdf_version)

    @property
    def interp_type(self) -> Interpolation1dType:
        """The 1D interpolation type."""
        return self._r.interp_type

    @property
    def psi_last(self) -> float:
        r"""The value of the last closed toroidal flux $\psi_{LCFS}$."""
        return self._r.psi_last

    @property
    def psip_last(self) -> float:
        r"""The value of the last closed toroidal flux $\psi_{p,LCFS}$."""
        return self._r.psip_last

    @property
    def psi_array(self) -> Array1:
        """The toroidal flux's values."""
        return self._r.get_array("psi_array")

    @property
    def psip_array(self) -> Array1:
        """The poloidal flux's values."""
        return self._r.get_array("psip_array")

    @property
    def g_array(self) -> Array1:
        """The poloidal plasma current $g$ values."""
        return self._r.get_array("g_array")

    @property
    def i_array(self) -> Array1:
        """The toroidal plasma current $I$ values."""
        return self._r.get_array("i_array")

File: crates/dexter-machine/tests/currents.rs
//! Test Currents functionality.

#![allow(unused_variables)]

use std::path::PathBuf;

use approx::assert_abs_diff_eq;
use dexter_machine::*;
use ndarray::Array1;

#[test]
fn lar_current() {
    let current = dbg!(LarCurrent::new());
    assert_eq!(current.psi_state(), FluxCoordinateState::Good);
    assert_eq!(current.psip_state(), FluxCoordinateState::Good);

    let mut acc = Accelerator::new();
    let psi = MagneticFlux::Toroidal(0.01);
    let psip = MagneticFlux::Poloidal(0.015);
    assert_abs_diff_eq!(current.eval_g(psi, &mut acc).unwrap(), 1.0);
    assert_abs_diff_eq!(current.eval_g(psip, &mut acc).unwrap(), 1.0);
    assert_abs_diff_eq!(current.eval_i(psi, &mut acc).unwrap(), 0.0);
    assert_abs_diff_eq!(current.eval_i(psip, &mut acc).unwrap(), 0.0);
    assert_abs_diff_eq!(current.eval_g_deriv(psi, &mut acc).unwrap(), 0.0);
    assert_abs_diff_eq!(current.eval_g_deriv(psip, &mut acc).unwrap(), 0.0);
    assert_abs_diff_eq!(current.eval_i_deriv(psi, &mut acc).unwrap(), 0.0);
    assert_abs_diff_eq!(current.eval_i_deriv(psip, &mut acc).unwrap(), 0.0);
}

#[test]
fn nc_current() {
    let path = PathBuf::from(extract::TEST_NETCDF_PATH);
    let typ = Interpolation1dType::Cubic;
    let builder = NcCurrentBuilder::new(&path, typ);
    let current = dbg!(builder.build().unwrap());
    assert_eq!(current.psi_state(), FluxCoordinateState::Good);
    assert_eq!(current.psip_state(), FluxCoordinateState::Good);

    let machine_type: MachineType = current.machine_type();
    let netcdf_version: semver::Version = current.netcdf_version();
    let path: PathBuf = current.path();
    let interp_type: Interpolation1dType = current.interp_type();
    let psi_state: FluxCoordinateState = current.psi_state();
    let psip_state: FluxCoordinateState = current.psip_state();
    let psi_last: f64 = current.psi_last().unwrap();
    let psip_last: f64 = current.psip_last().unwrap();
    let psi_array: Array1<f64> = current.psi_array().unwrap();
    let psip_array: Array1<f64> = current.psip_array().unwrap();
    let g_array: Array1<f64> = current.g_array();
    let i_array: Array1<f64> = current.i_array();

    let mut acc = Accelerator::new();
    let psi = MagneticFlux::Toroidal(0.01);
    let psip = MagneticFlux::Poloidal(0.015);
    let _: f64 = current.eval_g(psi, &mut acc).unwrap();
    let _: f64 = current.eval_g(psip, &mut acc).unwrap();
    let _: f64 = current.eval_i(psi, &mut acc).unwrap();
    let _: f64 = current.eval_i(psip, &mut acc).unwrap();
    let _: f64 = current.eval_i_deriv(psi, &mut acc).unwrap();
    let _: f64 = current.eval_i_deriv(psip, &mut acc).unwrap();
    let _: f64 = current.eval_i_deriv(psi, &mut acc).unwrap();
    let _: f64 = current.eval_i_deriv(psip, &mut acc).unwrap();
}
